web/hooks/pull_request.go

```diff
@@ -0,0 +1,198 @@
+		RemoteID: uint64(payload.GetSender().GetID()),
+		Provider: "github",
+	})
+	if err != nil {
+		wh.logger.Errorf("Failed to get user from database: %v", err)
+		return
+	}
+	reviewer, err := wh.db.GetEnrollmentByCourseAndUser(course.GetID(), user.GetID())
+	if err != nil {
+		wh.logger.Errorf("Failed to get reviewer from database: %v", err)
+		return
+	}
+
+	// Only if the review is from a course teacher, do we set the pull request to approved
+	if reviewer.IsTeacher() {
+		pullRequest.SetApproved()
+		wh.db.UpdatePullRequest(pullRequest)
+		wh.logger.Debugf("Pull request successfully approved for repository: %s", payload.GetRepo().GetFullName())
+	}
+}
+
+func (wh GitHubWebHook) handlePullRequestOpened(payload *github.PullRequestEvent) {
+	wh.logger.Debugf("Received pull request opened event for repository: %s, in organization: %s",
+		payload.GetRepo().GetName(), payload.GetOrganization().GetLogin())
+
+	repos, err := wh.db.GetRepositoriesWithIssues(&pb.Repository{RepositoryID: uint64(payload.GetRepo().GetID())})
+	if err != nil {
+		wh.logger.Errorf("Failed to get repository by remote ID %d from database: %v", payload.GetRepo().GetID(), err)
+		return
+	}
+	if len(repos) != 1 {
+		wh.logger.Debugf("Ignoring pull request opened event for unknown repository: %s", payload.GetRepo().GetFullName())
+		return
+	}
+	repo := repos[0]
+	if !repo.IsGroupRepo() {
+		wh.logger.Debugf("Ignoring pull request opened event for non-group repository: %s", payload.GetRepo().GetFullName())
+		return
+	}
+	wh.createPullRequest(payload, repo)
+}
+
+func (wh GitHubWebHook) handlePullRequestClosed(payload *github.PullRequestEvent) {
+	wh.logger.Debugf("Received pull request closed event for repository: %s, in organization: %s",
+		payload.GetRepo().GetName(), payload.GetOrganization().GetLogin())
+
+	if !payload.PullRequest.GetMerged() {
+		wh.logger.Debugf("Ignoring pull request closed event for non-merged pull request #%d, in %s",
+			payload.GetPullRequest().GetNumber(), payload.GetRepo().GetFullName())
+		return
+	}
+
+	pullRequest, err := wh.db.GetPullRequest(&pb.PullRequest{
+		ExternalRepositoryID: uint64(payload.GetRepo().GetID()),
+		Number:               uint64(payload.GetPullRequest().GetNumber()),
+	})
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			wh.logger.Debugf("Ignoring pull request closed event for non-managed pull request #%d, in %s",
+				payload.GetPullRequest().GetNumber(), payload.GetRepo().GetFullName())
+		} else {
+			wh.logger.Errorf("Failed to get pull request from database: %v", err)
+		}
+		return
+	}
+
+	if err := wh.db.HandleMergingPR(pullRequest); err != nil {
+		wh.logger.Errorf("Failed to delete pull request from database %v", err)
+	}
+	wh.logger.Debugf("Pull request successfully closed for repository: %s", payload.GetRepo().GetFullName())
+}
+
+// TODO(Meling): I think it makes sense to have this function here, instead of in assignments/pull_request.go,
```

Comment:
It is fine for now at least. Not sure where is the best place for it atm. But I do see that its only side-effect is updating the database. It is also reading the database. The payload is GitHub specific, so moving it elsewhere would pollute other parts with the GitHub stuff.

Reply:
You can remove the TODO comment.